### DISCUSSION TITLE
Allow setting mods in editor for test play

### DISCUSSION THREAD

**Discussion by LeNitrous** [ppy/osu#17396] [Ideas]
This is to allow testing of the map with specific mods quicker instead of the usual flow of exiting the editor and setting mods through song select.

Proof of Concept:

https://user-images.githubusercontent.com/20495991/159377165-8a1181f7-2cbf-4e47-a230-121d5170c6b7.mp4

I'm not fully convinced of the UX here but it works for now as a PoC.

**Comment by bdach** [comment]
really weird to have it in the editor rather than in the gameplay test screen itself. would like to think it would make more sense for it to be in the gameplay test screen (at the cost of not being a 10 second implementation).

**Reply by LeNitrous** [reply]
That may require restarting which might be a nuisance? Though I do agree it makes more sense for it to be there. We'll see once feedback arrives I suppose.

**Reply by bdach** [reply]
well it would have to work without restarting. hence the "wouldn't be a 10 second implementation" part.

**Reply by smoogipoo** [reply]
Eh. I'd say restarting is fine for now.

**Reply by peppy** [reply]
Any reason it can't be implemented as a panel in the `TestPlayerLoader` screen? Seems like the perfect place to put this for now. Have a panel showing the current mods and a button to open the overlay to change them.

**Reply by LeNitrous** [reply]
I don't think there is other than exposing `PlayerLoader`'s logic regarding the loading process of a `Player`.
EDIT: Might need consideration for Autoplay as it requires loading a `ReplayPlayer` for full playback controls.

**Reply by peppy** [reply]
Do keep in mind that seeking and what not should be possible even without autoplay turned on in the test mode. The currently implementation of test play is a bit of a placeholder.

**Reply by LeNitrous** [reply]
If this were the case then maybe only allow certain mods to be enabled in the overlay?

**Reply by peppy** [reply]
I'm not sure I follow. What is the reason for needing to disable certain mods?

**Reply by LeNitrous** [reply]
Some mods wouldn't make sense in test play. For instance, conversion mods like Target Practice and Difficulty Adjust. Automation mods like Cinema (and possibly Autoplay if the plan is to replicate stable where you can toggle autoplay on and off).

**Reply by bdach** [reply]
target practice/random/other conversion mods would *potentially* still make some degree of sense because while the beatmap is modified it's not like it's generated entirely without basing it on the real beatmap. random at least preserves distances between notes i believe and mostly randomizes angles.

**Reply by LeNitrous** [reply]
That's a fair argument. I'll leave this as a proposal for now until the editor has matured a bit more.

**Comment by peppy** [comment]
Can you explain when you would use this as a mapper? That will probably help shape the UX.

**Reply by LeNitrous** [reply]
To test maps with a specific combination when its intended use is for tournaments or when a mapper intends to apply a gimmick to a specific section of the map.

**Reply by hwabis** [reply]
As a player, I'd also use this to practice parts of a map without needing to start at the beginning, and at different speeds (with halftime mod). I know the intended use of the editor/testplaying is for mappers but just putting this here.

**Reply by bdach** [reply]
@hwabis that particular use case would be addressed by a special practice mod which has been proposed already.
